Feature task2

AlexLukashevich95 opened the pull request:
-changed jdbctemplate to jdbc
-added dml and ddl files

AlexLukashevich95 committed:
Feature-task2:
-changed jdbctemplate to jdbc
-added dml and ddl sqripts

AlexLukashevich95 committed:
Feature-task2:
-clear useless import

AlexLukashevich95 committed:
Feature-task2:
-delete comma in sql script

Evgnanii commented on database/ddl/create_tasktable.sql:
```diff
@@ -0,0 +1,8 @@
+-- Создание таблицы tasks
+CREATE TABLE tasks (
```
wer

Evgnanii replied:
rfherg

Evgnanii commented on database/ddl/create_tasktable.sql:
```diff
@@ -0,0 +1,8 @@
+    description TEXT,                -- Описание задачи
+    deadline DATE,                   -- Крайний срок выполнения задачи
+    status VARCHAR(50) NOT NULL     -- Статус задачи (например, 'Pending', 'In Progress', 'Completed')
+);
```
Убрать комменты. Рассмотреть другие варианты хранения даты. Например timestamp. Выяснить какой лучше и почему

Evgnanii commented on database/ddl/create_tasktable.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE tasks (
+    id SERIAL PRIMARY KEY,           -- Уникальный идентификатор задачи
+    name VARCHAR(255) NOT NULL,      -- Название задачи
+    description TEXT,                -- Описание задачи
```
Выяснить почему name varchar(255) а description TEXT. Подумать нужно ли для имени 255 символов

Evgnanii commented on database/dml/insert_tasks.sql:
```diff
@@ -0,0 +1,5 @@
+INSERT INTO tasks (name, description, deadline, status)
+VALUES
+    ('task 1', 'Implement the A feature', '2024-09-01', 'Done'),
```
изменить дату на другой тип

Evgnanii replied:
Вынести статусы в отдельную колонку

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
     public List<Task> findAll() {
-        String query = "SELECT * FROM tasks";
-        return jdbcTemplate.query(query, new BeanPropertyRowMapper<>(Task.class));
+        String query = "SELECT id,name,description,deadline,status FROM tasks";
```
вынести запросы в константы. Отформатировать их(добавить пробелы и отступы на след строкуhttps://medium.com/@candice173/best-practices-sql-formatting-85c2f4c874ec

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
-        String query = "SELECT * FROM tasks";
-        return jdbcTemplate.query(query, new BeanPropertyRowMapper<>(Task.class));
+        String query = "SELECT id,name,description,deadline,status FROM tasks";
+        List<Task> taskList = new ArrayList<Task>();
```
Не нужен дженерик в конструкторе

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
-        return jdbcTemplate.query(query, new BeanPropertyRowMapper<>(Task.class));
+        String query = "SELECT id,name,description,deadline,status FROM tasks";
+        List<Task> taskList = new ArrayList<Task>();
+        try (Connection con = dataSource.getConnection();
```
вынестри логику создания, закрытия, менеджмента сессии в отдельный Utility класс

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+                taskList.add(task);
+            }
+
+        } catch (SQLException e) {
```
сделать корректную обработку исключений. Тут должен выбрасываться кастомный эксепшн, который будет обрабатываться на уровне контроллера. На UI отобразить пользователю ошибку

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+        } catch (SQLException e) {
+            e.printStackTrace();
+        }
+        return taskList;
```
переименовать на tasks

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
 
     @Override
     public Task findById(Long id) {
         String query = "SELECT * FROM tasks WHERE id=?";
```
вынести в константу, добавить форматирование по правилам выше

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+                    System.out.println("No Employee found with id=" + id);
+                }
+            }
+        } catch (SQLException e) {
```
сделать корректную обратотку исключений

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
 
     @Override
     public Task save(Task task) {
         String query = "INSERT INTO tasks (name, description, deadline, status) VALUES (?, ?, ?, ?)";
```
вынести в константу, добавить форматирование

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
     @Override
     public Task save(Task task) {
         String query = "INSERT INTO tasks (name, description, deadline, status) VALUES (?, ?, ?, ?)";
-        jdbcTemplate.update(query, task.getName(), task.getDescription(), task.getDeadline(), task.getStatus());
```
подумать и реализовать механизм апдейта по частям, а не всех полей сразу

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
     public Task save(Task task) {
         String query = "INSERT INTO tasks (name, description, deadline, status) VALUES (?, ?, ?, ?)";
-        jdbcTemplate.update(query, task.getName(), task.getDescription(), task.getDeadline(), task.getStatus());
+        Connection con = null;
```
При выносе логики менеддмента сессии не забыть использовать try с ресурсами, чтобы не было как здесь

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+
+            ps.setString(1, task.getName());
+            ps.setString(2, task.getDescription());
+            ps.setDate(3, new java.sql.Date(task.getDeadline().getTime()));
```
рассмотреть другие варианты маппинга и работы с данными. java.sql.Date плохой класс

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+            } else {
+                System.out.println("task save failed with id=" + task.getId());
+            }
+        } catch (SQLException e) {
```
разобраться с обработкой ошибок

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+        } catch (SQLException e) {
+            if (con != null) {
+                try {
+                    con.rollback(); // Откатываем транзакцию при ошибке
```
если оставляешь комментарии, они должны быть там, где они действительно нужны будут другим разработчикам. Тут это не требуется. И комментарии оставляй на английском. Gpt тебе их генерит чтобы ты понял что он написал, не дублируй в чистовик

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
     public Task update(Task task) {
-        String query="update tasks SET name = ?, description = ?, deadline = ?, status = ? where id = ?";
-        jdbcTemplate.update(query, task.getName(), task.getDescription(), task.getDeadline(), task.getStatus(), task.getId());
+        String query = "update tasks SET name = ?, description = ?, deadline = ?, status = ? where id = ?";
```
вынести в константу. Сделать форматирование

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
-        jdbcTemplate.update(query, task.getName(), task.getDescription(), task.getDeadline(), task.getStatus(), task.getId());
+        String query = "update tasks SET name = ?, description = ?, deadline = ?, status = ? where id = ?";
+        Connection con = null;
+        PreparedStatement ps = null;
```
try с  ресурсами

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+            con = dataSource.getConnection();
+            ps = con.prepareStatement(query);
+
+            con.setAutoCommit(false);
```
зачем менять коммит мод? разобраться с этим и объяснить в чем необходимость. Если будешь убирать, то тоже спрошу почему убрал

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+
+            ps.setString(1, task.getName());
+            ps.setString(2, task.getDescription());
+            ps.setDate(3, new java.sql.Date(task.getDeadline().getTime()));
```
формат даты

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+        PreparedStatement ps = null;
+        try {
+            con = dataSource.getConnection();
+            ps = con.prepareStatement(query);
```
плохое навание переменных здесь и в других местах. Сделай connection,  prepareStatement/statement. Лучше больше букв, но пусть сразу будет понятно что это

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
     public void deleteById(Long id) {
         String query = "DELETE FROM tasks WHERE id=?";
-        jdbcTemplate.update(query, id);
+        Connection con = null;
```
try c ресурсами

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
 
     @Override
     public void deleteById(Long id) {
         String query = "DELETE FROM tasks WHERE id=?";
```
в константу. форматирование

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+        PreparedStatement ps = null;
+        try {
+            con = dataSource.getConnection();
+            ps = con.prepareStatement(query);
```
название переменных

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+            con = dataSource.getConnection();
+            ps = con.prepareStatement(query);
+
+            con.setAutoCommit(false);
```
автокоммит, зачем?

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+        } catch (SQLException e) {
+            if (con != null) {
+                try {
+                    con.rollback(); // Откатываем транзакцию при ошибке
```
комментарий лишний и на русском

Evgnanii commented on src/main/java/com/alex/tasktable/repository/impl/TaskRepositoryImpl.java:
```diff
@@ -3,48 +3,205 @@
+                try {
+                    con.rollback(); // Откатываем транзакцию при ошибке
+                    System.out.println("Transaction rolled back due to error: " + e.getMessage());
+                } catch (SQLException rollbackEx) {
```
обработка ошибок

AlexLukashevich95 committed:
Feature-task2:
-changed date format
-added Status table and made changes connected to it
-added custom exception
-added constants for queries
-made connection module
-added yml properties file
-changed dataSource to basicDataSource

AlexLukashevich95 committed:
Feature-task2:
-added select field to taskeditform

Evgnanii commented on database/ddl/create_tasktable.sql:
```diff
@@ -0,0 +1,15 @@
+);
+
+-- Создание таблицы tasks
+CREATE TABLE stats (
```
Будем переименовывать на statuses, stats это характеристики

Evgnanii commented on src/main/java/com/alex/tasktable/config/YamlConfig.java:
```diff
@@ -0,0 +1,43 @@
+public class YamlConfig {
+    public static DataSourceProperties dataSourceProperties;
+
+    static {
```
будем использовать спринговые стуки чтобы считать проперти
